Mathlib/RingTheory/PowerSeries/WeierstrassPreparation.lean

```diff
@@ -435,4 +454,191 @@ theorem weierstrassMod_zero_left [IsAdicComplete (IsLocalRing.maximalIdeal A) A]
+    have := H.coeff_f_sub_r_mem hi
+    rwa [map_sub, coeff_X_pow, if_neg hi.ne, zero_sub, neg_mem_iff, Polynomial.coeff_coe] at this
+  · have := congr($(H.2) * ↑(H.isUnit_of_map_ne_zero hg).unit⁻¹)
+    rw [add_mul, mul_assoc, IsUnit.mul_val_inv, mul_one, ← sub_eq_iff_eq_add] at this
+    simp_rw [← this, f, Polynomial.coe_sub, Polynomial.coe_pow, Polynomial.coe_X, sub_mul]
+
+theorem IsWeierstrassFactorization.isWeierstrassDivision
+    {g : A⟦X⟧} {f : A[X]} {h : A⟦X⟧} (H : g.IsWeierstrassFactorization f h) :
+    (X ^ (g.map (IsLocalRing.residue A)).order.toNat).IsWeierstrassDivision g ↑H.isUnit.unit⁻¹
+      (Polynomial.X ^ (g.map (IsLocalRing.residue A)).order.toNat - f) := by
+  set n := (g.map (IsLocalRing.residue A)).order.toNat with hn
+  constructor
+  · refine (Polynomial.degree_sub_lt ?_ (Polynomial.monic_X_pow n).ne_zero ?_).trans_eq (by simpa)
+    · simp_rw [H.degree_eq_coe_lift_order_map, Polynomial.degree_X_pow, n,
+        ENat.lift_eq_toNat_of_lt_top]
+    · rw [(Polynomial.monic_X_pow n).leadingCoeff, H.isDistinguishedAt.monic.leadingCoeff]
+  · simp_rw [H.eq_mul, mul_assoc, IsUnit.mul_val_inv, mul_one, Polynomial.coe_sub,
+      Polynomial.coe_pow, Polynomial.coe_X, add_sub_cancel]
+
+section IsAdicComplete
+
+variable [IsAdicComplete (IsLocalRing.maximalIdeal A) A] {g : A⟦X⟧}
+  (hg : g.map (IsLocalRing.residue A) ≠ 0)
+include hg
+
+/-- **Weierstrass preparation theorem** ([washington_cyclotomic], Theorem 7.3):
+let `g` be a power series over a complete local ring,
+such that the image of `g` in the residue field is not zero. Then there exists a distinguished
+polynomial `f` and a power series `h` which is a unit, such that `g = f * h`. -/
+theorem exists_isWeierstrassFactorization : ∃ f h, g.IsWeierstrassFactorization f h := by
+  obtain ⟨q, r, H⟩ :=
+    (X ^ (g.map (IsLocalRing.residue A)).order.toNat).exists_isWeierstrassDivision hg
+  exact ⟨_, _, H.isWeierstrassFactorization hg⟩
+
+/-- The `f` in Weierstrass preparation theorem. -/
+noncomputable def weierstrassDistinguished : A[X] :=
+  (g.exists_isWeierstrassFactorization hg).choose
+
+/-- The `h` in Weierstrass preparation theorem. -/
+noncomputable def weierstrassUnit : A⟦X⟧ :=
+  (g.exists_isWeierstrassFactorization hg).choose_spec.choose
+
+theorem isWeierstrassFactorization_weierstrassDistinguished_weierstrassUnit :
+    g.IsWeierstrassFactorization (g.weierstrassDistinguished hg) (g.weierstrassUnit hg) :=
+  (g.exists_isWeierstrassFactorization hg).choose_spec.choose_spec
+
+theorem isDistinguishedAt_weierstrassDistinguished :
+    (g.weierstrassDistinguished hg).IsDistinguishedAt (IsLocalRing.maximalIdeal A) :=
+  (g.exists_isWeierstrassFactorization hg).choose_spec.choose_spec.isDistinguishedAt
+
+theorem isUnit_weierstrassUnit :
+    IsUnit (g.weierstrassUnit hg) :=
+  (g.exists_isWeierstrassFactorization hg).choose_spec.choose_spec.isUnit
+
+theorem eq_weierstrassDistinguished_mul_weierstrassUnit :
```

Comment:
A `simp` lemma, inverting the order of the `mul`?

Reply:
I don't think so, because in application this result is used to rewrite `g` to the right hand side.